Update mill to 0.12.4

nightscape opened the pull request:
@ckipp01 this one was a little bit of a PITA, so it would be nice to get it merged soonish 😉 
It adds compatibility with Mill 0.12, but still retains compatibility with previous versions.
Only the `reconcileRanges` test doesn't work, not sure what changed there...

ckipp01 commented on build.sc:
```diff
@@ -102,25 +102,31 @@ trait ItestCross extends Cross.Module[String] with MillIntegrationTestModule {
+        "reconciledRange" -> "verify",
+        "cyclical" -> "checkManifest"
+      ).filter {
+          // The reconciledRange test doesn't work under 0.12 atm
```
So this was due to what I believe is a bug in coursier. I created an issue for it in https://github.com/coursier/coursier/issues/2682 but never got a response about what is actually happening. Maybe this is fixed with the new changes in coursier, but I'd need to dig deeper to better understand if that's the case. Right now you can go ahead and just ignore this and I'll try to look deeper at what's happening.

ckipp01 replied:
I dug a bit deeper and created https://github.com/com-lihaoyi/mill/issues/4165.

ckipp01 commented on build.sc:
```diff
@@ -102,25 +102,31 @@ trait ItestCross extends Cross.Module[String] with MillIntegrationTestModule {
     T {
+      val env = if(millTestVersion() >= "0.12")
+        Map(
+          "COURSIER_REPOSITORIES" -> s"central sonatype:releases ivy:file://${T.dest.toString.replaceFirst("testInvocations", "test")}/ivyRepo/local/[organisation]/[module]/[revision]/[type]s/[artifact].[ext]"
```
So is this needed now for 0.12 versions?

nightscape replied:
Yes, this was the only way I found to make Mill / Coursier pick up the local dev repo.
In Mill 0.12+ there is a different way to integration-test Mill plugins, with which this hack/workaround would probably go away. But that doesn't work for 0.10 and 0.11, so I guess it's best to live with the workaround until usage of Mill < 0.12 dies out and one can completely switch to the new approach.

ckipp01 reviewed (approved):
Thanks!

nightscape commented:
@ckipp01 I hope/think I fixed the remaining issues:

- Updated from Java 8 to 11 (Mill 0.12.0-RC1 dropped support for Java 8)
- Fixed the version typo in the `reconciledRange` test
- Changed the way the tests discover the directory containing `manifests.json`

Could you approve GitHub Actions to run the tests again?

nightscape commented:
I fixed all logged style errors and force-pushed the results.
Locally I still get an error like this with no further explanation:
```
domain.fix Something unexpected happened running Scalafix
```
I hope this doesn't happen in the GitHub Actions.
@ckipp01 could you trigger another run?

nightscape commented:
@ckipp01 the same error happened as on my local machine.
I updated the `mill-scalafix` plugin and ran the tests locally using [act](https://github.com/nektos/act).
I got up to the `submit-dependency-graph` step, which I think should fail locally, as I don't have the required GitHub credentials.
So hopefully this is the final try 😉

nightscape commented:
@ckipp01 all green, ready to merge from my side 👍

ckipp01 reviewed (approved):
Thanks a lot for your work on this @nightscape. LGTM

nightscape commented:
@ckipp01 looks like the release process has 0.12 problems...
We could
- build the project using Mill 0.11, but still publish artifacts for 0.10, 0.11 and 0.12
- update mill-ci-release to support Mill 0.12
- change the build to publish without mill-ci-release

ckipp01 commented:
Yes, I see that. I think maybe for now we should just use 0.11.x to publish, then try to update mill-ci-release to work with 0.12.x.
